File: kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define DEFAULT_ITER 200
#define EPSILON 0.001
#define ERROR_MESSAGE "An Error Has Occurred"

typedef struct{
    double* centroid;
    double* prevCentroid;
    double** dataPointList;
    int dataPointsSize;
}CLUSTER;

void clearClusterDataPointList(CLUSTER *cluster){
    cluster -> dataPointsSize = 0;
}

void updateClusterCentroid(CLUSTER *cluster, int d){
    int i = 0;
    int assignedPoints = cluster-> dataPointsSize;

    if (assignedPoints == 0){
        printf(ERROR_MESSAGE);
        return;
    }
  
    for (i = 0; i < d; i++){
        double sum=0.0;
        int j;
        for (j = 0; j < assignedPoints; j++){
            sum += cluster -> dataPointList[j][i];
        }
        cluster->prevCentroid[i] = cluster->centroid[i];
        cluster -> centroid[i] = sum/(assignedPoints);
    }
}

void addDataPointToCluster(CLUSTER *cluster, double* dataPoint){
    cluster -> dataPointList[cluster->dataPointsSize] = dataPoint;
    cluster -> dataPointsSize++;
}

double calcDistance(double* vector1, double* vector2, int d){
    double sum = 0.0;
    int i=0;
    for (i = 0; i < d; i++){
        sum += pow((vector1[i] - vector2[i]), 2);
    }
    return sqrt(sum);
}

int hasClustersConverged(int K, int d, CLUSTER* clusters){
    int i;
    for (i = 0; i < K; i++){
        if (calcDistance(clusters[i].centroid, clusters[i].prevCentroid, d) >= EPSILON){
            return 0;
        }
    }
    return 1;
}

void assignClusterToDataPoint(CLUSTER* clusters, double** dataPoints, int N, int K, int d){
    int pointIndex;
    for (pointIndex = 0;  pointIndex < N; pointIndex++){
        int clusterIndex;
        int minClusterIdx = 0;
        double minDistance = calcDistance(dataPoints[pointIndex], clusters[0].centroid, d);
        for (clusterIndex = 1; clusterIndex < K; clusterIndex++){
            double distance = calcDistance(dataPoints[pointIndex], clusters[clusterIndex].centroid, d);
            if (distance < minDistance){
                minDistance = distance;
                minClusterIdx = clusterIndex;
            }
        
        }
        addDataPointToCluster(&clusters[minClusterIdx], dataPoints[pointIndex]);
    }
}

void freeDataPoints(double** dataPoints, int N){
    int i;
    for (i = 0; i < N; i++){
        free(dataPoints[i]);
    }
    free(dataPoints);
}

void freeClusters(CLUSTER* clusters, int K) {
    int i;
    for (i = 0; i < K; i++) {
        free(clusters[i].centroid);
        free(clusters[i].prevCentroid);
        free(clusters[i].dataPointList);
    }
    free(clusters);
}

CLUSTER* initializeClusters(int K, int N, int d, double** dataPoints){
    CLUSTER* clusters = (CLUSTER*)malloc(K * sizeof(CLUSTER));
    int i = 0;
    if (!clusters) {
        printf(ERROR_MESSAGE);
        return NULL;
    }

    for(i=0; i<K; i++){
        int j = 0;
        clusters[i].centroid = (double*)malloc(d * sizeof(double));
        clusters[i].prevCentroid =(double*) malloc(d * sizeof(double));
        clusters[i].dataPointList = (double**)malloc(N *sizeof(double*));
        clusters[i].dataPointsSize = 0;
        if (!clusters[i].centroid || !clusters[i].prevCentroid || !clusters[i].dataPointList) {
            printf(ERROR_MESSAGE);
            freeClusters(clusters, i+1);
            return NULL;
        }

        for (j = 0; j < d; j++) {
            clusters[i].centroid[j] = dataPoints[i][j];
            clusters[i].prevCentroid[j] = dataPoints[i][j];
        }
    }
    return clusters;
}

void printClusters(int K, int d, CLUSTER* clusters){
    int clusterIdx;
    for (clusterIdx = 0; clusterIdx < K; clusterIdx++){
        int i;
        for (i = 0; i < d; i++){
            printf("%.4f", clusters[clusterIdx].centroid[i]);
            if (i<d-1){
                printf(",");
            } 
        }
        if (clusterIdx < K-1){
            printf("\n");
        }
    }
}

int solveKMeanAlgo(int K, int N, int d, int iter, double** dataPoints){
    int counter = 0;
    CLUSTER* clusters = initializeClusters(K, N, d, dataPoints);
    if (!clusters){
        return 1;
    }

    for (counter = 0; counter < iter; counter++){
        int i=0;
        assignClusterToDataPoint(clusters, dataPoints, N, K, d);
    
        for (i = 0; i < K; i++){
            updateClusterCentroid(&clusters[i], d);
            clearClusterDataPointList(&clusters[i]);
        }

        if (hasClustersConverged(K, d, clusters)){
            break;
        }
    }
    printClusters(K, d, clusters);
    freeClusters(clusters, K);   
    return 0;
}

double** allocateDataPointsMemory(int d, int N){
    double** dataPoints;
    int i = 0;

    dataPoints = (double**)calloc(N, sizeof(double*));
    if (!dataPoints){
        printf(ERROR_MESSAGE);
        return NULL;
    }
    
    for (i = 0; i < N; i++){
        dataPoints[i] = (double*)calloc(d, sizeof(double));
        if (!dataPoints[i]){
            freeDataPoints(dataPoints, i);
            printf(ERROR_MESSAGE);
            return NULL;
        }
    }
    return dataPoints;  
}

double** retrieveDataPoints(int N, int d){
    int i = 0;
    int j = 0;
    double** dataPoints = allocateDataPointsMemory(d, N);

    if (!dataPoints){
        return NULL;
    }

    for (i = 0; i < N; i++){
        for (j = 0; j < d; j++){
            if (scanf("%lf,", &(dataPoints[i][j]))!=1){
                scanf("%lf\n", &(dataPoints[i][j]));
            }    
        }
    }

    return dataPoints;
}

int isInteger(const char *str) {
    if (!*str)
        return 0;

    while (*str) {
        char c = *str;
        if (c < '0' || c > '9')
        {
           return 0;
        }
        str++;
    }
    return 1;
}

int* retrieveAndValidateInputs(int argc, char** argv) {
    int N = 0;
    int K = 0;
    int d = 0;
    int iter = DEFAULT_ITER;
    int* inputs = NULL;

    if (argc < 4) {
        printf(ERROR_MESSAGE);
        return NULL;
    }
    
    if (!isInteger(argv[2]) || atoi(argv[2]) <= 1) {
        printf("Invalid number of points!");
        return NULL;
    }
    N = atoi(argv[2]);

    if (!isInteger(argv[1]) || atoi(argv[1]) <= 1 || atoi(argv[1]) >= N) {
        printf("Invalid number of clusters!");
        return NULL;
    }
    K = atoi(argv[1]);

    if (!isInteger(argv[3]) || atoi(argv[3]) < 1) {
        printf("Invalid dimension of point!");
        return NULL;
    }
    d = atoi(argv[3]);

    if (argc > 4) {
        iter = atoi(argv[4]);
        if (!isInteger(argv[4]) || iter <= 1 || iter >= 1000) {
            printf("Invalid maximum iteration!");
            return NULL;
        }
    }

    inputs = (int*)malloc(4 * sizeof(int));
    if (!inputs) {
        printf(ERROR_MESSAGE);
        return NULL;
    }

    inputs[0] = K;
    inputs[1] = N;
    inputs[2] = d;
    inputs[3] = iter;

    return inputs;
}

int main(int argc, char** argv){
    double** dataPoints = NULL;
    int res = 0;
    int* inputs = retrieveAndValidateInputs(argc, argv);

    if (!inputs){
        return 1;
    }
    dataPoints = retrieveDataPoints(inputs[1], inputs[2]);
    if (!dataPoints){
        free(inputs);
        return 1;
    }

    res = solveKMeanAlgo(inputs[0], inputs[1], inputs[2], inputs[3], dataPoints);

    freeDataPoints(dataPoints, inputs[1]);
    free(inputs);
    return res;
}

File: kmeans.py
import sys
import numpy as np

DEAFULT_ITER = 200
DECIMAL_DIGITS = 4
ERROR_MESSAGE = "An Error Has Occurred"
epsilon = 0.001

class Cluster:
    def __init__(self, centroid:np.array):
        self.centroid = centroid
        self.prev_centroid = None
        self.data_points_list = []

    def assign_data_point(self, data_point):
        self.data_points_list.append(data_point)
    
    def clear_data_points_list(self):
        self.data_points_list.clear()

    def update_centroid(self):
        self.prev_centroid = self.centroid
        self.centroid = np.mean(self.data_points_list, axis=0)

    def print_centroid(self):
        d = len(self.centroid)
        for i in range(0, d):
            value = format(self.centroid[i], f'.{DECIMAL_DIGITS}f')
            if i < d-1:
                print(value, end=",")
            else:
                print(value)            

class KMeansAlgo:
    def __init__(self, K, N, d, iter, input_data):
        self.K = K
        self.N = N
        self.d = d
        self.iter = iter
        self.input_data = input_data

    def _has_clusters_converged(self, clusters):
        for cluster in clusters:
            if self._get_distance(cluster.prev_centroid, cluster.centroid) >= epsilon:
                return False
        return True


    def _get_distance(self, vector1, vector2):
        return np.linalg.norm(vector1-vector2)

    def _assign_cluster_to_datapoint(self, clusters, datapoints):
        for data_point in datapoints:
            distances = np.array([self._get_distance(data_point, cluster.centroid) for cluster in clusters])
            closest_cluster_idx = np.argmin(distances)
            clusters[closest_cluster_idx].assign_data_point(data_point)

    def _get_datapoints_from_file(self):
        with open(self.input_data, 'r') as file:
            datapoints = []
            counter = 0
            while True or counter<self.N:
                line = file.readline()
                if not line:
                    break
                point = np.fromstring(line, dtype=float, sep=',')
                datapoints.append(point)
                counter+= 1
            return datapoints

    def solve_kmeans_algo(self):
        data_points = self._get_datapoints_from_file()
        clusters = [Cluster(data_point) for data_point in data_points[:self.K]]
        for counter in range(0, self.iter):
            self._assign_cluster_to_datapoint(clusters, data_points)
            for cluster in clusters:
                cluster.update_centroid()
                cluster.clear_data_points_list()

            if self._has_clusters_converged(clusters):
                break
        return clusters


def print_clusters(clusters, d):
    for cluster in clusters:
        cluster.print_centroid()
        
def retrieve_argv_and_validate():
    if len(sys.argv) < 5:
        print(ERROR_MESSAGE)
        return None
    
    if not sys.argv[2].isnumeric() or int(sys.argv[2])<=1:
        print("Invalid number of points!")
        return None
    N = int(sys.argv[2])

    K = sys.argv[1]
    if not K.isnumeric() or int(K) <=1 or int(K) >= N:
        print("Invalid number of clusters!")
        return None
    K = int(K)
    
    d = sys.argv[3]
    if not d.isnumeric() or int(d)<1:
        print("Invalid dimension of point!")
    d = int(d)

    iter = DEAFULT_ITER
    fileName = ""

    if len(sys.argv) > 5:
        iter = sys.argv[4]
        if not iter.isnumeric() or int(iter) <= 1 or int(iter) >= 1000:
                print("Invalid maximum iteration!")
                return None
        iter = int(iter)
        fileName = sys.argv[5]    
    else:
        fileName = sys.argv[4]

    return K,N,d,iter,fileName

        
if __name__ == "__main__":
    parameters = retrieve_argv_and_validate()
    if parameters != None:
       kmeansAlgo = KMeansAlgo(*parameters)
       print_clusters(kmeansAlgo.solve_kmeans_algo(), parameters[2])
       


